torch/fx/passes/infra/pass_manager.py

```diff
@@ -0,0 +1,269 @@
+import inspect
+from functools import wraps
+from typing import Any, Callable, Dict, List, Tuple
+
+import torch.nn as nn
+from torch.fx._compatibility import compatibility
+
+@compatibility(is_backward_compatible=False)
+def inplace_wrapper(fn: Callable) -> Callable:
+    """
+    Convenience wrapper for passes which modify an object inplace. This
+    wrapper makes them return a PassResult containing the modified object and
+    True for the "modified" flag.
+
+    Args:
+        fn (Callable[Module, Any])
+
+    Returns:
+        wrapped_fn (Callable[Module, PassResult])
+    """
+
+    @wraps(fn)
+    def wrapped_fn(gm):
+        fn(gm)
+        return gm
+
+    return wrapped_fn
+
+def _validate_pass_schedule_constraint(
+    constraint: Callable[[Callable, Callable], bool], passes: List[Callable]
+) -> None:
+    for i, a in enumerate(passes):
+        for j, b in enumerate(passes[i + 1 :]):
+            if constraint(a, b):
+                continue
+            raise RuntimeError(
+                f"pass schedule constraint violated. Expected {a} before {b}"
+                f" but found {a} at index {i} and {b} at index{j} in pass"
+                f" list."
+            )
+
+@compatibility(is_backward_compatible=False)
+def topological_sort_passes(
+    passes: List[Callable], constraints: List[Callable]
+) -> Tuple[List[Callable], bool]:
+    """
+    Args
+        passes: Passes that we are ordering
+        constraints: Constraints applied on these passes
+
+    Returns
+        A sorted list of callables and a boolean of if a circular dependency
+        existed
+    """
+    if len(constraints) == 0:
+        return passes, False
+
+    # Construct a graph
+    graph: Dict[Callable, List[Callable]] = {}
+    visited: Dict[Callable, Any] = {}
+    res: List[Callable] = []
+    for p in passes:
+        graph[p] = []
+        visited[p] = None
+
+    circular_dependency = False
+    for i, a in enumerate(passes):
+        for j, b in enumerate(passes):
+            if i == j:
+                continue
+
+            for constraint in constraints:
+                if not constraint(a, b):
+                    if b in graph[a]:
+                        circular_dependency = True
+                    graph[b].append(a)
+
+    time = 1
+
+    # Topologically sort the graph
+    def topological_sort_util(graph, p):
```

Comment:
IIUC, this is a DFS based topological sort, and recursively calls itself. 
In production system, we tends avoid recursively call, as you can get into really deep calling stack if one is not careful...

I find the in_degree map traversal more understandable for topological sort, and it's probably shorter to write... 

See topo_sort in fuser_utils.py.

Reply:
Updated the code to use the iterative version of topological sort! The code isn't exactly shorter due to the issue with breaking cycles, but the actual sorting part is shorter and more readable!